src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,779 @@
+                }
+            };
+
+            sorter.sort(0, leafValueCount);
+
+            // write leaf rowID -> orig index
+            scratchOut2.reset();
+
+            // iterate in row ID order to get the row ID index for the given value order index
+            // place into an array to be written as packed ints
+            for (int x = 0; x < leafValueCount; x++)
+                orderIndex[rowIDAndIndexes[x].valueOrderIndex] = x;
+
+            LeafOrderMap.write(orderIndex, leafValueCount, maxPointsInLeafNode - 1, scratchOut2);
+
+            treeOutput.writeVInt(scratchOut2.getPosition());
+            treeOutput.writeBytes(scratchOut2.getBytes(), 0, scratchOut2.getPosition());
+
+            callback.writeLeafDocs(rowIDAndIndexes, 0, leafValueCount);
+
+            writeCommonPrefix(scratchOut, commonPrefixLength, leafValues);
+
+            scratchBytesRef1.length = bytesPerValue;
+            scratchBytesRef1.bytes = leafValues;
+
+            IntFunction<BytesRef> packedValues = (i) -> {
+                scratchBytesRef1.offset = bytesPerValue * i;
+                return scratchBytesRef1;
+            };
+            if (DEBUG)
+                valuesInOrderAndBounds(leafValueCount,
+                                       ArrayUtil.copyOfSubArray(leafValues, 0, bytesPerValue),
+                                       ArrayUtil.copyOfSubArray(leafValues, (leafValueCount - 1) * bytesPerValue, leafValueCount * bytesPerValue),
+                                       packedValues,
+                                       leafRowIDs);
+
+            writeLeafBlockPackedValues(scratchOut, commonPrefixLength, leafValueCount, packedValues);
+
+            treeOutput.writeBytes(scratchOut.getBytes(), 0, scratchOut.getPosition());
+
+            scratchOut.reset();
+        }
+
+        private void checkMaxLeafNodeCount(int numLeaves)
+        {
+            if ((1 + bytesPerValue) * (long) numLeaves > ArrayUtil.MAX_ARRAY_LENGTH)
+            {
+                throw new IllegalStateException("too many nodes; increase maxPointsInLeafNode (currently " + maxPointsInLeafNode + ") and reindex");
+            }
+        }
+
+        private void writeCommonPrefix(DataOutput treeOutput, int commonPrefixLength, byte[] packedValue) throws IOException
+        {
+            treeOutput.writeVInt(commonPrefixLength);
+            if (commonPrefixLength > 0)
+                treeOutput.writeBytes(packedValue, 0, commonPrefixLength);
+        }
+
+        private void writeLeafBlockPackedValues(DataOutput out, int commonPrefixLength, int count, IntFunction<BytesRef> packedValues) throws IOException
+        {
+            if (commonPrefixLength == bytesPerValue)
+            {
+                // all values in this block are equal
+                out.writeByte((byte) -1);
+            }
+            else
+            {
+                assert commonPrefixLength < bytesPerValue;
```

Comment:
nit: Throw in a quick error message if we actually keep this I guess...

Reply:
I've removed it. The `commonPrefixLength` is calculated just before this method call and really can't fail this assertion. I find these assertions pointless and possibly dangerous because they cast doubt on things that really can't be wrong.